python_modules/dagster/dagster/_core/definitions/dependency.py

```diff
@@ -486,22 +487,25 @@ def from_dict(cls, dict_repr: Mapping[str, Any]) -> "NodeHandle":
         return NodeHandle(name=dict_repr["name"], parent=parent)
 
 
+T_OptionalNodeHandle = TypeVar("T_OptionalNodeHandle", bound=Optional[NodeHandle])
+
+
 @dagster_model(checked=False)
-class NodeInputHandle:
+class NodeInputHandle(Generic[T_OptionalNodeHandle]):
```

Comment:
what does the type var set us up for?

Reply:
so the difference here over just `Optional[NodeHandle]` is that the type checker can know when its not `None` if it knew it was not `None` at construction (within that scope at least)?

Reply:
exactly

Reply:
which empirically does appear to save needing to add some `check.not_none`s

Reply:
maybe worth highlighting with a comment as its pretty subtle, I had to stare at this for a while to get it

Reply:
👍 👍 . added one above the `TypeVar`. let me know if you think it could be clearer